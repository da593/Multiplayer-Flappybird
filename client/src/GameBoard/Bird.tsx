import React,{useContext, useEffect, useRef} from 'react';
import { DimensionContext } from 'hooks/DimensionsContext';
import { BoxCoordinates } from '@flappyblock/shared';

interface Props {
    birdCoords: BoxCoordinates;
    isSelf: boolean;
}

export function Bird({isSelf, birdCoords}:Props) {

    const dimensions = useContext(DimensionContext);
    
    const canvasRef = useRef<HTMLCanvasElement | null>(null);

    const draw = (context:CanvasRenderingContext2D) => {
        if (dimensions) {
            context.clearRect(0, 0, context.canvas.width, context.canvas.height)
            context.fillStyle = isSelf ? "red" : "orange";
            context.fillRect(50, birdCoords.topLeft.y, dimensions.BIRD_WIDTH, dimensions.BIRD_WIDTH);
        }
    }

    useEffect(() => {
        const canvas = canvasRef.current;
        if (canvas ) {
            const context = canvas.getContext("2d");
            if (context) {
                draw(context);
            }
        }
    },[birdCoords])

    return (
        <canvas className="canvas-item" width={dimensions ? dimensions.GAME_WIDTH : undefined} height={dimensions ? dimensions.GAME_HEIGHT : undefined} ref={canvasRef}/>
    )
}