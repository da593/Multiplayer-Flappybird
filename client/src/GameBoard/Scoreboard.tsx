
import React from 'react';

interface Props {
    score: number;
}

export function Scoreboard({score}:Props) {

    return (
        <p className="score">{score}</p>
    )
}