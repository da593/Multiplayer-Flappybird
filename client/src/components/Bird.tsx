import React,{useEffect, useRef} from 'react';
import { BirdCoordinates, BoardSize } from './BoardGame';
import * as consts from "components/GameConstants";

export type BirdProps = BoardSize & BirdCoordinates
export function Bird(props:BirdProps) {

    const canvasRef = useRef<HTMLCanvasElement | null>(null);

    const draw = (context:CanvasRenderingContext2D) => {
        context.clearRect(0, 0, context.canvas.width, context.canvas.height)
        context.fillStyle = "red";
        context.fillRect(50, props.yBirdLoc, consts.BIRD_WIDTH, consts.BIRD_WIDTH);

    }

    useEffect(() => {
        const canvas = canvasRef.current;
        if (canvas ) {
            const context = canvas.getContext("2d");
            if (context) {
                draw(context);
            }
        }
    },[props.yBirdLoc])

    return (
        <canvas id="bird" width={props.width} height={props.height} ref={canvasRef}/>
    )
}