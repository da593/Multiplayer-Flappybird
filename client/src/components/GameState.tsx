import React, {useState,useEffect} from "react";
import { BoardGame } from "./BoardGame";
import * as consts from "components/GameConstants";

export interface Coordinates {
    x: number;
    y: number;
}

export interface BoxCoordinates {
    topLeft: Coordinates;
    topRight: Coordinates;
    botLeft: Coordinates;
    botRight: Coordinates;
}

export function GameState() {
    const [xPipeLoc,setXPipeLoc] = useState(consts.GAME_WIDTH + consts.PIPE_WIDTH);
    const [yGapLoc,setYGapLoc] = useState(randomizeYGapLocation());
    const [gapCoords,setGapCoords] = useState<BoxCoordinates>({topLeft: {x: 0, y:0}, topRight: {x: 0, y:0}, botLeft: {x: 0, y:0}, botRight: {x: 0, y:0}});

    const [yBirdLoc,setYBirdLoc] = useState(consts.GAP_HEIGHT / 3);
    const [birdCoords,setBirdCoords] = useState<BoxCoordinates>({topLeft: {x: 0, y:0}, topRight: {x: 0, y:0}, botLeft: {x: 0, y:0}, botRight: {x: 0, y:0}});

    const [hasCollided,setHasCollided] = useState(false);
    const [hasKeyClicked,setHasKeyClicked] = useState(false);

    function randomizeYGapLocation():number {
        return Math.floor(Math.random() * (consts.GAME_WIDTH - consts.GAP_HEIGHT));
    }

    const detectCollision = () => {
        if (birdCoords.botLeft.y >= consts.GAME_HEIGHT) {
            setHasCollided(true);
        }
    }   

    const animatePipe = (frameCount:number) => {
        if (frameCount == 0) {
            setYGapLoc(randomizeYGapLocation());
            setXPipeLoc(consts.GAME_WIDTH + consts.PIPE_WIDTH);
        }
        else {
            setXPipeLoc(xPipeLoc - frameCount);
        }  
    }

    const animateBird = (frameCount:number,isEvent=false) => {
        if (hasKeyClicked == true) {
            setYBirdLoc(yBirdLoc - 40);
            setHasKeyClicked(false);
        }
        else if (frameCount == 0) {
            setYBirdLoc(consts.GAP_HEIGHT / 3)
        }
        else {
            setYBirdLoc(yBirdLoc + frameCount);
        }
    }

    const handleClick = (e:KeyboardEvent) => {
        if (e.key === "w") {
            setHasKeyClicked(true);
        }
    }    
    useEffect(() => {
        let raf:number;
        let frameCount = 0;
        const render = () => {
            if (!hasCollided) {
                frameCount == consts.GAME_WIDTH + consts.PIPE_WIDTH ? frameCount = 0 : frameCount++;
                animatePipe(frameCount)
                raf = window.requestAnimationFrame(render);
            }
        }
        render();
        return () => {
            window.cancelAnimationFrame(raf);
        }
    },[hasCollided])

    useEffect(() => {
        window.addEventListener("keydown",handleClick)
        let raf:number;
        let frameCount = 0;
        const render = () => {
            if (!hasCollided) {
                frameCount++
                animateBird(frameCount);
                raf = window.requestAnimationFrame(render);
            }
        }
        render();
        return () => {
            window.cancelAnimationFrame(raf);
        }
    },[hasCollided,hasKeyClicked])
    
    useEffect(() => {
        const gapTopLeft:Coordinates = {x:xPipeLoc - consts.PIPE_WIDTH, y:yGapLoc - consts.GAP_HEIGHT};
        const gapTopRight:Coordinates = {x:xPipeLoc, y:yGapLoc - consts.GAP_HEIGHT};
        const gapBotLeft:Coordinates = {x:xPipeLoc - consts.PIPE_WIDTH, y:yGapLoc};
        const gapBotRight:Coordinates = {x:xPipeLoc, y:yGapLoc};
        setGapCoords({topLeft:gapTopLeft,topRight: gapTopRight, botLeft: gapBotLeft, botRight: gapBotRight})

        const birdTopLeft:Coordinates = {x: consts.BIRD_X_LOCATION - consts.BIRD_WIDTH, y: yBirdLoc - consts.BIRD_WIDTH};
        const birdTopRight:Coordinates = {x: consts.BIRD_X_LOCATION, y: yBirdLoc - consts.BIRD_WIDTH};
        const birdBotLeft:Coordinates = {x: consts.BIRD_X_LOCATION - consts.BIRD_WIDTH, y: yBirdLoc};
        const birdBotRight:Coordinates = {x: consts.BIRD_X_LOCATION, y: yBirdLoc};

        setBirdCoords({topLeft:birdTopLeft,topRight:birdTopRight, botLeft:birdBotLeft, botRight:birdBotRight})
    },[xPipeLoc,yBirdLoc])

    useEffect(() => {
        detectCollision()
    },[gapCoords,birdCoords])


    return (
        <BoardGame width={consts.GAME_WIDTH} height={consts.GAME_HEIGHT} xPipeLoc={xPipeLoc} yGapLoc={yGapLoc} yBirdLoc={yBirdLoc}/>
    )
}