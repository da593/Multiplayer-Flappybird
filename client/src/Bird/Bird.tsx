import React,{useEffect, useRef} from 'react';
import * as consts from "GameState/constants";
import { BirdCoordProps, GameSizeProps } from 'GameState/types';

type BirdProps = BirdCoordProps & GameSizeProps;
export function Bird(props:BirdProps) {

    const canvasRef = useRef<HTMLCanvasElement | null>(null);

    const draw = (context:CanvasRenderingContext2D) => {
        context.clearRect(0, 0, context.canvas.width, context.canvas.height)
        context.fillStyle = "red";
        context.fillRect(50, props.birdCoords.topLeft.y, consts.BIRD_WIDTH, consts.BIRD_WIDTH);

    }

    useEffect(() => {
        const canvas = canvasRef.current;
        if (canvas ) {
            const context = canvas.getContext("2d");
            if (context) {
                draw(context);
            }
        }
    },[props])

    return (
        <canvas id="bird" width={props.width} height={props.height} ref={canvasRef}/>
    )
}