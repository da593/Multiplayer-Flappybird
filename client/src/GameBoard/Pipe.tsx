import React,{useEffect, useRef} from 'react';
import { GapCoordProps } from 'GameState/types';
import { GAME_DIMENSIONS } from 'GameState/constants';

type PipeProps = GapCoordProps;
export function Pipe(props: PipeProps) {
  
    const canvasRef = useRef<HTMLCanvasElement | null>(null)

    const draw = (canvas:HTMLCanvasElement,context:CanvasRenderingContext2D) => {
        context.clearRect(0, 0, context.canvas.width, context.canvas.height)

        context.fillStyle = "green";
        context.fillRect(props.gapCoords.topLeft.x, 0, GAME_DIMENSIONS.PIPE_WIDTH, props.gapCoords.topLeft.y);
        context.fillRect(props.gapCoords.topLeft.x, props.gapCoords.botLeft.y, GAME_DIMENSIONS.PIPE_WIDTH, canvas.height)

        context.fillStyle= "#87CEEB"; 
        context.fillRect(props.gapCoords.topLeft.x, props.gapCoords.topLeft.y , GAME_DIMENSIONS.PIPE_WIDTH, GAME_DIMENSIONS.GAP_HEIGHT);
        
    }



    useEffect(() => {
        const canvas = canvasRef.current;
        if (canvas ) {
            const context = canvas.getContext("2d");
            if (context) {
                draw(canvas,context);
            }
        }
    },[props])


    return (
        <canvas className="canvas-item" width={GAME_DIMENSIONS.GAME_WIDTH} height={GAME_DIMENSIONS.GAME_HEIGHT} ref={canvasRef} />
    )
}