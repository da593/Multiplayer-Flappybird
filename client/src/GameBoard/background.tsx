import React,{useEffect, useRef} from 'react';
import { GameSizeProps } from 'GameState/types';


export function BoardBackground(props:GameSizeProps) {

    const canvasRef = useRef<HTMLCanvasElement>(null);
    
    useEffect(() => {
        if (canvasRef.current ) {
            const canvas = canvasRef.current;
            const context = canvas.getContext("2d");
            if (context) {
                context.fillStyle = "#87CEEB";
                context.fillRect(0, 0, props.width,props.height);
            } 

        }
    },[])

    return (
        <canvas id="background-layer" width={props.width} height={props.height} ref={canvasRef}/>
    )
}
