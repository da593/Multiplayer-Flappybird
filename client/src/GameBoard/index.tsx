import React from 'react';
import { BoardBackground } from './background';
import { Scoreboard } from 'GameBoard/Scoreboard';
import { Pipe } from './Pipe';
import { Bird } from './Bird';
import { DimensionContext } from 'hooks/DimensionsContext';
import { useLoaderData } from 'react-router-dom';
import { Dimensions_I, GameState } from '@flappyblock/shared';

interface Props extends GameState {
    hasStarted: boolean;
}

export function BoardGame({player, pipe, hasStarted}: Props)  {
    const dimensions = useLoaderData() as Dimensions_I;
    return (
        <div className={pipe.hasCollided || !hasStarted ? "gameboard opaque" : "gameboard"}>
            <DimensionContext.Provider value={dimensions}>
                <Scoreboard score={player.score}/>
                <Pipe gapCoords={pipe.gapCoords}/>
                <Bird  birdCoords={player.birdCoords}/>
                <BoardBackground/> 
            </DimensionContext.Provider>
        </div>
    )
}

