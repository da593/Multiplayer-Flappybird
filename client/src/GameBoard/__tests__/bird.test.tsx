import React  from 'react';
import { Bird } from 'GameBoard/Bird';
import renderer from 'react-test-renderer';
import { BoxCoordinates } from 'GameState/types';
import { GAME_DIMENSIONS } from 'GameState/constants';

const birdAtCeiling:BoxCoordinates = {
    topLeft:{x: GAME_DIMENSIONS.BIRD_X_LOCATION, y: 0}, 
    topRight:{x: GAME_DIMENSIONS.BIRD_X_LOCATION + GAME_DIMENSIONS.BIRD_WIDTH, y: 0}, 
    botLeft:{x: GAME_DIMENSIONS.BIRD_X_LOCATION, y: GAME_DIMENSIONS.BIRD_WIDTH}, 
    botRight:{x: GAME_DIMENSIONS.BIRD_X_LOCATION + GAME_DIMENSIONS.BIRD_WIDTH, y:  GAME_DIMENSIONS.BIRD_WIDTH}
};

const birdAtY:BoxCoordinates = {
    topLeft:{x: GAME_DIMENSIONS.BIRD_X_LOCATION, y: 50}, 
    topRight:{x: GAME_DIMENSIONS.BIRD_X_LOCATION + GAME_DIMENSIONS.BIRD_WIDTH, y: 50}, 
    botLeft:{x: GAME_DIMENSIONS.BIRD_X_LOCATION, y: GAME_DIMENSIONS.BIRD_WIDTH}, 
    botRight:{x: GAME_DIMENSIONS.BIRD_X_LOCATION + GAME_DIMENSIONS.BIRD_WIDTH, y:  GAME_DIMENSIONS.BIRD_WIDTH}
};

test("render bird given props", () => {
    const component = renderer.create(<Bird birdCoords={birdAtCeiling}/>)
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot()

    
})
