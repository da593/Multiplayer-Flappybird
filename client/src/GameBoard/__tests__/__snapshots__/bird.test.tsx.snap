// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render bird given props 1`] = `
<canvas
  className="canvas-item"
  height={640}
  width={540}
/>
`;
